purekit/src/main/kotlin/com/virgilsecurity/purekit/pure/Pure.java

```diff
@@ -106,114 +126,65 @@ public Pure(PureContext context) throws CryptoException, PureException {
-     * @return AuthResult with PureGrant and encrypted PureGrant
-     * @throws ProtocolHttpException FIXME
-     * @throws ProtocolException FIXME
-     * @throws CryptoException FIXME
+     * Authenticates user.
+     *
+     * @param userId User Id.
+     * @param password Password.
+     * @param sessionId Optional sessionId which will be present in PureGrant.
+     *
+     * @return AuthResult with PureGrant and encrypted PureGrant.
+     *
+     * @throws ProtocolException Thrown if an error from the PHE service has been parsed
+     * successfully.
+     * @throws ProtocolHttpException Thrown if an error from the PHE service has NOT been parsed
+     * successfully. Represents a regular HTTP exception with code and message.
+     * @throws PureException If provided password is invalid or - please, see
+     * {@link PureStorage#selectUser(String)} PureException doc.
      */
-    public AuthResult authenticateUser(String userId, String password, String sessionId) throws ProtocolHttpException, ProtocolException, Exception {
-        if (userId == null || userId.isEmpty()) {
-            throw new NullPointerException();
-        }
-        if (password == null || password.isEmpty()) {
-            throw new NullPointerException();
-        }
+    public AuthResult authenticateUser(String userId, String password, String sessionId)
+        throws ProtocolException, ProtocolHttpException, PureException {
 
-        byte[] passwordHash = this.crypto.computeHash(password.getBytes(), HashAlgorithm.SHA512);
+        ValidateUtils.checkNullOrEmpty(new HashMap<String, String>() {{
+            put(userId, "userId");
+            put(password, "password");
+        }});
 
-        UserRecord userRecord = this.storage.selectUser(userId);
+        byte[] phek = computePheKey(userId, password);
 
-        PheClient client = this.getClient(userRecord.getPheRecordVersion());
+        byte[] uskData = cipher.decrypt(storage.selectUser(userId).getEncryptedUsk(), phek);
 
-        byte[] pheVerifyRequest = client.createVerifyPasswordRequest(passwordHash, userRecord.getPheRecord());
-
-        PurekitProtos.VerifyPasswordRequest request = PurekitProtos.VerifyPasswordRequest.newBuilder()
-                .setVersion(userRecord.getPheRecordVersion())
-                .setRequest(ByteString.copyFrom(pheVerifyRequest))
-                .build();
-
-        PurekitProtos.VerifyPasswordResponse response = this.pheClient.verifyPassword(request);
-
-        byte[] phek = client.checkResponseAndDecrypt(passwordHash, userRecord.getPheRecord(), response.getResponse().toByteArray());
-
-        if (phek.length == 0) {
-            throw new PureException(PureException.ErrorCode.INVALID_PASSWORD);
+        VirgilKeyPair ukp;
+        try {
+            ukp = this.crypto.importPrivateKey(uskData);
+        } catch (CryptoException exception) {
+            throw new IllegalStateException("Exception while importing key. This should not"
```

Comment:
uskData depends on what is in the storage. Any error that appears while working with external systems or their data MUST NOT be RuntimeException